### PR TITLE
First pass at JWT auto-deletion flag

### PR DESCRIPTION
**Pull request by tdsacilowski**
Fixes https://github.com/hashicorp/vault/issues/11968

Still need to write tests.

### CODE REVIEW

**Comment by calvn** [comment]
Are there potential (re-)authentication issues that could arise if we don't delete a stale JWT?

File: website/content/docs/agent/autoauth/methods/jwt.mdx [hunk 1]
```diff
@@ -17,3 +17,7 @@ JWT to perform a reauthentication.
 - `path` `(string: required)` - The path to the JWT file
 
 - `role` `(string: required)` - The role to authenticate against on Vault
+
+- `remove_jwt_after_reading` `(bool: optional, defaults to true)` -
```
Comment (calvn) [inline website/content/docs/agent/autoauth/methods/jwt.mdx:21]: Can we flip this around and have it be `keep_after_reading` or something along those lines? This way the default value is also the zero value (`false`).

Comment (tdsacilowski) [reply website/content/docs/agent/autoauth/methods/jwt.mdx:21]: I thought about that but I thought it more prudent to retain the current default behavior and opt-out if needed. I'd assume that in _most_ cases, the default behavior is more appropriate. This change is mostly to allow for using Kubernetes bound service account tokens with JWT auth. In this use case, Kubernetes automatically rotates the JWT based on a configurable TTL.

Comment (calvn) [reply website/content/docs/agent/autoauth/methods/jwt.mdx:21]: `keep_after_reading` with it being defaulted `false` should still retain the current behavior though. Note that we're flipping the naming and also the default so keeping the files would still be an opt-in. We'd only _not_ delete the file if this gets set to `true` explicitly.

**Comment by tdsacilowski** [comment]
> Are there potential (re-)authentication issues that could arise if we don't delete a stale JWT?

I wouldn't think so. JWTs _should_ have an expiration set so if re-auth is attempted with an expired JWT auth should fail. In this scenario, the underlying system that is responsible for delivering an updated JWT (e.g. Kubernetes) should have provided a newer JWT. But I may be missing some of the finer points here.

File: website/content/docs/agent/autoauth/methods/jwt.mdx (hunk 1 above)
Comment (tdsacilowski) [reply website/content/docs/agent/autoauth/methods/jwt.mdx:21]: Ah ok, I see what you mean. I was following the same model that the AppRole auto-auth was using but I can change this around if it makes more sense.

**Comment by calvn** [comment]
It looks like `ingressToken()` [relies on the absence](https://github.com/hashicorp/vault/pull/11969/files#diff-91e3ede29e58c49d358b61b7c7b14faa3585be554040c0fb1a6f4603d7212b7aR170-R172) on the file to short-circuit from the function, so on non-kubernetes workflows in which someone opts in for this, I could see Wgent doing the opposite -- reading the file every 500ms. We do end up comparing the value of this token with the existing one so it's mostly benign, but it does result in more filesystem reads.

**Comment by aelbarkani** [comment]
@arianvp since this PR was not merged, did you find a workaround ?

File: website/content/docs/agent/autoauth/methods/jwt.mdx (hunk 1 above)
Comment (tomhjp) [reply website/content/docs/agent/autoauth/methods/jwt.mdx:21]: I actually like `remove_jwt_after_reading` for its consistency with AppRole's option: https://www.vaultproject.io/docs/agent/autoauth/methods/approle#remove_secret_id_file_after_reading

IMO the bar to break from this pattern needs to be higher than making it terser. If it wasn't for the prior art, I would 100% agree with both the flip + naming comments above, but they're minor nits and they're secondary to preventing confusion for consumers.

File: command/agent/auth/jwt/jwt.go [hunk 2]
```diff
@@ -73,6 +76,15 @@ func NewJWTAuthMethod(conf *auth.AuthConfig) (auth.AuthMethod, error) {
...
 	}
 
+	removeJWTAfterReadingRaw, ok := conf.Config["remove_jwt_after_reading"]
+	if ok {
+		removeJWTAfterReading, err := parseutil.ParseBool(removeJWTAfterReadingRaw)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing 'remove_jwt_after_reading' value: %w", err)
+		}
+		j.removeJWTAfterReading = removeJWTAfterReading
+	}
```
Comment (tomhjp) [inline command/agent/auth/jwt/jwt.go:86]:
```suggestion
	if removeJWTAfterReadingRaw, ok := conf.Config["remove_jwt_after_reading"]; ok {
		removeJWTAfterReading, err := parseutil.ParseBool(removeJWTAfterReadingRaw)
		if err != nil {
			return nil, fmt.Errorf("error parsing 'remove_jwt_after_reading' value: %w", err)
		}
		j.removeJWTAfterReading = removeJWTAfterReading
	} else {
		j.removeJWTAfterReading = true
	}
```

Comment (tomhjp) [reply command/agent/auth/jwt/jwt.go:86]: Seems like the diff doesn't display very well, but the main comment is to default the value to true to maintain existing behaviour.

**Review by tomhjp — COMMENTED** [review]
Taking another look as there has been some renewed interest in this. @tdsacilowski I'm happy to take this on if you'd prefer not to pick this up again.

File: command/agent/auth/jwt/jwt.go [hunk 3]
```diff
@@ -73,14 +75,32 @@ func NewJWTAuthMethod(conf *auth.AuthConfig) (auth.AuthMethod, error) {
...
 
+	if removeJWTAfterReadingRaw, ok := conf.Config["remove_jwt_after_reading"]; ok {
+		removeJWTAfterReading, err := parseutil.ParseBool(removeJWTAfterReadingRaw)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing 'remove_jwt_after_reading' value: %w", err)
+		}
+		j.removeJWTAfterReading = removeJWTAfterReading
+	} else {
+		// Default to true for maximum backwards compatibility.
+		j.removeJWTAfterReading = true
```
Comment (tvoran) [inline command/agent/auth/jwt/jwt.go:86]: It looks like this is defaulting to `true` in the initial declaration of `j`. Do we need it both places?

**Review by tvoran — APPROVED** [review]
I'm ambivalent about the option name.

File: command/agent/auth/jwt/jwt.go (hunk 3 above)
Comment (tomhjp) [reply command/agent/auth/jwt/jwt.go]: Oops, good spot. I lost that in the formatting diff noise.

**Pull request merged**